### DISCUSSION TITLE
How to write a custom component with both lua function and highlighting group?

### DISCUSSION THREAD

**Discussion by linrongbin16** [nvim-lualine/lualine.nvim#1178] [Q&A]
Hi,

I want to create a custom component with lua function (since it's flexiable) to produce the text content, and render it with a highlighting group (for example 'Function').

how should I do it?

for now I configure it with:

```lua

local SCROLL_BAR = { "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█" }

local function ScrollBar()
    local curr_line = vim.api.nvim_win_get_cursor(0)[1]
    local lines = vim.api.nvim_buf_line_count(0)
    local i = math.floor((curr_line - 1) / lines * #SCROLL_BAR) + 1
    return string.rep(SCROLL_BAR[i], 2)
end

require('lualine').setup({
  ...
  sections = {
    ...
    lualine_z = { 'progress', ScrollBar },
  }
})   
```

It looks like:

<img width="412" alt="image" src="https://github.com/nvim-lualine/lualine.nvim/assets/6496887/dccd6bc2-4c66-456b-ba59-69cc5e5aaec8">

I want to render the text with some color, so I try to configure it with:

```lua

local SCROLL_BAR = { "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█" }

local function ScrollBar()
    local curr_line = vim.api.nvim_win_get_cursor(0)[1]
    local lines = vim.api.nvim_buf_line_count(0)
    local i = math.floor((curr_line - 1) / lines * #SCROLL_BAR) + 1
    return "%#String#" .. string.rep(SCROLL_BAR[i], 2)  -- I add '%#String#' here
end
```

But it seems breaks the background color of `lualine_z` sections:

<img width="562" alt="image" src="https://github.com/nvim-lualine/lualine.nvim/assets/6496887/4ed0b543-dec7-4b46-9181-e4e3a396af8f">

**Comment by linrongbin16** [comment]
I finally configure it with:

```lua
        lualine_z = {
            "progress",
            {
                ScrollBar,
                color = function()
                    local mode = vim.o.termguicolors and "gui" or "cterm"
                    local code = vim.fn.synIDattr(
                        vim.fn.synIDtrans(vim.fn.hlID("Special")),
                        "fg",
                        mode
                    )
                    return { fg = code, gui = "bold" }
                end,
            },
        }

```

And it's working now:

<img width="202" alt="image" src="https://github.com/nvim-lualine/lualine.nvim/assets/6496887/27cac4f8-cd7b-4ce3-a0b2-d4b52d95cc4c">

**Reply by maxigaz** [reply]
Thank you for the solution! I've been looking for a way to replace Lualine's native progress indicator with a bar icon, and also a way to change the colour of a component (not an entire section). You helped me achieve both. Cheers!
